src/mrinufft/io/utils.py

```diff
@@ -35,3 +35,34 @@ def add_phase_to_kspace_with_shifts(kspace_data, kspace_loc, normalized_shifts):
     phi = np.sum(kspace_loc * normalized_shifts, axis=-1)
     phase = np.exp(-2 * np.pi * 1j * phi)
     return kspace_data * phase
+
+
+def siemens_quat_to_rot_mat(quat):
+    """
+    Calculate the rotation matrix from Siemens Twix quaternion.
+    """
+    a = quat[1]
+    b = quat[2]
+    c = quat[3]
+    d = quat[0]
+
+    R = np.zeros((4, 4))
+    
+    R[0,1] = 1.0 - 2.0 * (b * b + c * c)
+    R[0,0] = 2.0 * (a * b - c * d)
+    R[0,2] = 2.0 * (a * c + b * d)
+
+    R[1,1] = 2.0 * (a * b + c * d)
+    R[1,0] = 1.0 - 2.0 * (a * a + c * c)
+    R[1,2] = 2.0 * (b * c - a * d)
+
+    R[2,1] = 2.0 * (a * c - b * d)
+    R[2,0] = 2.0 * (b * c + a * d)
+    R[2,2] = 1.0 - 2.0 * (a * a + b * b)
+
+    if (np.linalg.det(R[:3, :3]) < 0):
+        R[2] = -R[2]
+        
+    R[-1,-1] = 1
```

Comment:
Why do it here rather than before the determinant check ?

Reply:
Saving compute xD. determinant of 3x3 is faster i presume than 4x4 :P